src/python/pants/backend/go/util_rules/build_opts.py

```diff
@@ -76,8 +99,24 @@ async def go_extract_build_options_from_target(
     if cgo_enabled is None:
         cgo_enabled = golang.cgo_enabled
 
+    # Extract the `with_race_detector` value for this target.
+    with_race_detector: bool | None = None
+    if go_test_subsystem.force_race:
+        with_race_detector = True
+    if test_target_fields is not None and test_target_fields.test_race.value is not None:
+        with_race_detector = test_target_fields.test_race.value
+    if with_race_detector is None:
+        if target_fields is not None and target_fields.race.value is not None:
+            with_race_detector = target_fields.race.value
+    if with_race_detector is None:
+        if go_mod_target_fields.race is not None:
+            with_race_detector = go_mod_target_fields.race.value
+    if with_race_detector is None:
+        with_race_detector = False
```

Comment:
This may be valuable to extract into a helper function so that you can unit test. Lots of conditions here

Reply:
Also how come not using `elif`?

Reply:
Each of those `if with_race_detector is None` clauses have an inner `if` clause that may not be triggered. Thus `with_race_detector` could continue to be `None`. `elif` would have different semantics.

Reply:
Simplified the code by introducing a `_first_non_none_value` helper. Added tests as well of the various combinations.